### ISSUE TITLE
Pointer type is parsed as LHS of multiply expression in `is` context

### ISSUE THREAD

**Issue by jhinder** [dotnet/roslyn#70049]
**Version Used**: VS 2022 17.8 Preview 2 (4.8.0-2.23429.7)

**Steps to Reproduce**:
Inspect the syntax tree of the following method.
```csharp
unsafe void M(int* a)
{
    _ = a is long*;
    _ = a is long* v;
}
```

**Expected Behavior**: The type is parsed as a pointer type even though pointer types cannot participate in pattern matching (error CS8521).

**Actual Behavior**:
Both `is` expressions are parsed as a constant pattern with a multiply expression with `long` as the LHS (CS1525 "Invalid expression term 'long'") and either:
1. a missing RHS (CS1525 "Invalid expression term ';'")
2. `v` as the RHS -- an attempt at a variable declaration pattern, but in this case it's parsed as an identifier (CS0103 "The name 'v' does not exist in the current context").

---
I don't know the details of the C# spec to say whether the current parse is the 'correct' one or whether my interpretation is equally valid. But when I was playing around with the code in the IDE I had assumed that any type would be accepted at the start of a type pattern, even if it was found to be invalid when bound. But I did not expect to see that parse tree.

In either case both interpretations are not valid programs/lead to the compiler emitting error diagnostics, so changing the interpretation would not really be a breaking change, would it? It would, however, lead to better diagnostics (for example CS8521 is not emitted right now despite it saying more clearly what is wrong).

**Comment by jaredpar** [comment]
> The type is parsed as a pointer type even though pointer types cannot participate in pattern matching (error CS8521).

Do not necessarily agree that is the expected behavior. Consider the following syntax is legal in patterns

```csharp
if (o is x * 13) { 
  ...
}
```

Given that pointers are illegal as types in patterns (there is no way to perform a test), it seems reasonable that we choose to parse for the valid syntax here. 

The spec is a bit ambiguous as it just says we supprot `type` but doesn't specify if pointers are / aren't allowed. 

@333fred does this sound reasonable?

**Comment by 333fred** [comment]
> The spec is a bit ambiguous as it just says we supprot type but doesn't specify if pointers are / aren't allowed.

I don't think this is ambiguous. We _did_ just specify patterns in the [v7 spec](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/patterns.md), and it explicitly says a `type` is expected there. In `unsafe` contexts, a `type` includes pointer types. Regardless of whether there is a semantic error, it seems clear to me that it should parse as a type, not as a multiplication statement. I'll also ping @MadsTorgersen and @BillWagner for their spec interpretations, in case they differ from my reading.

**Comment by MadsTorgersen** [comment]
Let me first say that the standard only talks about correct programs, so it doesn't technically have anything to say on which error to give here.

However, the parse as a multiplication expression seems odd to me, because it is invalid according to the grammar: `long` is a keyword (https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/lexical-structure.md#644-keywords), not an identifier.

Parsing as a _declaration_pattern_ on the other hand is at least grammatically correct, and fails a later semantic check. So that seems "closer to the truth" and a more useful error.

But I want to stress that this is entirely subjective. As far as I know we have no hard and fast rules on deciding which error to prioritize.

**Comment by BillWagner** [comment]
I'd agree with @333fred 's analysis. When in an unsafe context, pointer types are types.

That said, @MadsTorgersen makes the most important point here: The standard doesn't allow the code in the sample. The roslyn compiler doesn't allow the code in the sample. Ideally, the implementation should pick the "best" diagnostic to help a developer resolve the error. That's a subjective call. My opinion would be to parse it as a pointer type. I wouldn't make a big argument about it though.
